First pull request for generation of prog_to_run file

michaeldwong opened the pull request:
Grammar:

Implemented alugrammar.lalrpop that generates AST nodes based on ast.rs. The grammar matches the .alu files and currently parses on all available .alu files. In addition to the specific tree nodes, ast.rs also has the generation code for the helper functions (mux2, mux3, rel_op, etc.). The AST traversal and code generation will be in the next pr. As mentioned in the last meeting, OptHeader is the optional header on top of the .alu files that passes in the pipeline stage number, alu number within that stage, and the prefix name to ast.rs for the proper name and hole generation — its purpose will be a lot more straightforward in the next pr.
 
Tests:

In lib.rs, tests from test_druzhba.rs and test_grammar.rs are called. test_druzhba.rs contains code similar to code to program_to_run_akv.rs. It initializes ALUs, pipeline stages, and the pipeline and feeds a few PHVs into it. Based on the hardcoded hole variables, muxes, and ALU behaviors, assertions were made on what the resulting values of the output PHVs will be. test_grammar.rs tests the grammar file, alugrammar.lalrpop and uses the generated parser against hardcoded strings that are common patterns in the .alu files. The test_stateful_alus and test_stateless_alus tests are the most important since they verify that the Chipmunk .alu files can be parsed successfully. 

Misc:

Added build.rs and updated Cargo.toml to work with LALRPOP processing. Added new instructions to README.md. Added all of the .alu files from the example_alus directory in Chipmunk. Made slight changes to the existing druzhba code for better coding style (such as changing mutable reference to immutable reference in alu.rs) and to mitigate compiler warnings.

michaeldwong committed:
ALU grammer from config branch, changes to previous druzhba code, test

aatish17varma committed:
Printing current tick for clarity

aatish17varma reviewed (commented):
Really nice testing

anirudhSK reviewed (changes requested):
Overall, this looks great. A couple of comments:

1. Split up the final test for test_grammar.rs into multiple smaller tests.

2. Document unused variables with either a comment or an attribute (e.g., like gcc's unused variable attribute).

3. Add a few tests to compare result of parsing against a known parse tree.

4. Move code from prog_to_run_akv.rs into test_druzhba.rs and delete prog_to_run_akv.rs.

michaeldwong committed:
Finished test_grammar tests

michaeldwong committed:
Fixed ast test. Deleted program_to_run_AKV, slight changes

michaeldwong committed:
Documentation for unused vars

michaeldwong committed:
Resolved few compiler warnings

michaeldwong committed:
Added new test to test_druzhba

michaeldwong commented:
1. Done.
2. Done.
3. Finished one test — can add 1 or 2 more, they'd just be tedious to write
4. Deleted program_to_run_AKV.rs and moved contents into prog_to_run.rs which is called by main. I didn't move that test into test_druzhba.rs because they're almost identical so I made a new test that verifies that that stateful ALU output is successfully written to the phv

michaeldwong committed:
Changed test names
